src/conn-management.ts

```diff
@@ -0,0 +1,158 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import * as utils from "./utils";
+import { IDeferred } from "./utils";
+import { MatrixClient } from "./client";
+import { ISyncStateData } from "./sync";
+import { Method } from "./http-api";
+import { SyncState } from "./sync";
+
+export type Callback = (newState: SyncState, data?: ISyncStateData) => void;
+
+/**
+ * ConnectionManagement is a class which handles keep-alives and invokes a callback when the connection
+ * is alive/dead. Uses /versions as a keep-alive endpoint.
+ */
+export class ConnectionManagement {
```

Comment:
This class is deeply flawed and provides questionable value.

1. `pokeKeepAlive` is only ever called with an arg of `0` so the param is useless as is the internal default value of 2000 + rand
2. Unless the `online` event fires (not available in RN) this class does nothing. `retryImmediately` will immediately return false unless `startKeepAlives` has already ran (via `online` firing)
3. The callback is only fired on error, not when the error goes away, so this cannot be used for "connection management"

Given a /sync failing 3 times already fires the sync error, this class does not provide any real gains beyond maybe improving the latency on a server error once the window fires an `online` event. We also never listen for the `offline` event wtf?

Reply:
Yes